java/arcs/android/sdk/host/ArcHostHelper.kt

```diff
@@ -0,0 +1,394 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.android.sdk.host
+
+import android.app.Service
+import android.content.ComponentName
+import android.content.Context
+import android.content.Intent
+import android.content.pm.ServiceInfo
+import android.os.Bundle
+import android.os.Handler
+import android.os.Parcelable
+import android.os.ResultReceiver
+import androidx.annotation.VisibleForTesting
+import arcs.android.host.parcelables.ActualParcelable
+import arcs.android.host.parcelables.ParcelableParticleIdentifier
+import arcs.android.host.parcelables.ParcelablePlanPartition
+import arcs.android.host.parcelables.toParcelable
+import arcs.core.host.ArcHost
+import arcs.core.host.ParticleIdentifier
+import arcs.core.host.PlanPartition
+import kotlinx.coroutines.CancellableContinuation
+import kotlinx.coroutines.CoroutineName
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlin.coroutines.CoroutineContext
+import kotlin.reflect.KClass
+
+/**
+ * Tool which can be used by [ArcHost]s to handle [Intent] based API calls, as well as
+ * send API calls.
+ *
+ * ## Example Usage:
+ *
+ * ```kotlin
+ * class MyService : Service() {
+ *     private val myHelper: ArcHostHelper by lazy {
+ *         ArcHostHelper(MyArcHost())
+ *     }
+ *
+ *     class MyArcHost : AbstractArcHost() {
+ *         override suspend fun onStartArc(plan: PlanPartition) {
+ *             // ...
+ *         }
+ *     }
+ *
+ *     override fun onStartCommand(intent: Intent?, flags: Int, startId: Int) {
+ *         val result = super.onStartCommand(intent, flags, startId)
+ *         myHelper.onStartCommand(intent)
+ *         return result
+ *     }
+ * }
+ * ```
+ */
+class ArcHostHelper(
+    private val service: Service,
+    private val arcHost: ArcHost,
+    private val coroutineContext: CoroutineContext = Dispatchers.IO
+) {
+
+    private val job = Job() + coroutineContext + CoroutineName("ArcHostHelper")
+
+    /**
+     * Determines whether or not the given [Intent] represents a call to an [ArcHost] and invokes
+     * the appropriate interface methods.
+     */
+    fun onStartCommand(intent: Intent?) = CoroutineScope(job).launch {
+        onStartCommandSuspendable(intent)
+    }
+
+    @VisibleForTesting
+    suspend fun onStartCommandSuspendable(intent: Intent?) {
+        // Ignore other actions
+        if (intent?.action?.startsWith(
+                ARC_HOST_INTENT
+            ) != true
+        ) return
```

Comment:
I feel like this could be a lot more readable.

```kotlin
val action = intent?.action ?: return
if (!action.startsWith(ARC_HOST_INTENT)) return
```